### PR TITLE
[macos] Hack around mono problematic facade issue with an exclusion list

### PR DESCRIPTION
**Pull request by chamons**
- generateExclusionLish.sh is a total hack and will be replaced later with nuget-hash-extractor

### CODE REVIEW

File: tools/mmp/driver.cs [hunk 1]
```diff
@@ -442,6 +442,29 @@ static void Main2 (string [] args)
...
+				}
+				if (referencesToFix.Count > 0) {
+					string facadeRoot = string.Format ("/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/{0}/Facades/", IsUnifiedMobile ? "Xamarin.Mac" : "4.5");
+
+					FixReferences (asm => referencesToFix.Contains (asm), asm => {
+						string replacement = Path.Combine (facadeRoot, Path.GetFileName (asm));
+						ErrorHelper.Warning (1407, string.Format ("'{0}' is a facade mono has declared problematic and refuses to load. Replacing reference with '{1}'.", asm, replacement));
+						return replacement;
+					});
+              	}
```
Comment (spouliot) [inline tools/mmp/driver.cs]: ^ spaces/tabs issue ?

File: tools/mmp/driver.cs [hunk 2]
```diff
@@ -517,6 +540,30 @@ static void Main2 (string [] args)
...
+		{
+			List<ExclusionItem> exclusionList = new List<ExclusionItem> ();
+			using (var exclusionStream = typeof (Driver).Assembly.GetManifestResourceStream ("monoExclusionList.txt")) {
+				if (exclusionStream == null) {
+					ErrorHelper.Warning (100, "Internal Warning \"Unable to load mono exclusion list.\" .Please file a bug report with a test case (http://bugzilla.xamarin.com).");
+					return exclusionList;
+				}
+				using (var sr = new StreamReader (exclusionStream)) {
+					while (!sr.EndOfStream) {
+						string [] lineParts = sr.ReadLine().Split (new char [] { ':' });
```
Comment (spouliot) [inline tools/mmp/driver.cs]: minor: space before `()`

File: tools/mmp/Makefile [hunk 3]
```diff
@@ -131,8 +131,8 @@ LOCAL_MMP = \
 	Mono.Cecil.dll \
 	Mono.Cecil.Mdb.dll \
 
-mmp.exe: Makefile $(MONO_CECIL_DLL) $(MONO_CECIL_MDB_DLL) $(mmp_sources) config config_mobile Info.plist.tmpl $(tuner_sources) $(linker_resources)
-	$(Q_MCS) $(SYSTEM_MCS) -unsafe -out:mmp.exe $(DEFINES) -r:$(MONO_CECIL_DLL) -r:$(MONO_CECIL_MDB_DLL) -r:Mono.Security.dll  -resource:config -resource:config_mobile -resource:machine.4_5.config -resource:Info.plist.tmpl $(linker_resources:%=-resource:%) $(tuner_sources) $(mmp_sources)
+mmp.exe: Makefile $(MONO_CECIL_DLL) $(MONO_CECIL_MDB_DLL) $(mmp_sources) config config_mobile Info.plist.tmpl $(tuner_sources) $(linker_resources) monoExclusionList.txt
+	$(Q_MCS) $(SYSTEM_MCS) -unsafe -out:mmp.exe $(DEFINES) -r:$(MONO_CECIL_DLL) -r:$(MONO_CECIL_MDB_DLL) -r:Mono.Security.dll -resource:monoExclusionList.txt -resource:config -resource:config_mobile -resource:machine.4_5.config -resource:Info.plist.tmpl $(linker_resources:%=-resource:%) $(tuner_sources) $(mmp_sources)
```
Comment (spouliot) [inline tools/mmp/Makefile]: I don't like the hard coding into mmp, even as a resource. Fine for a temporary hack but that should not get into master (or future release branches)

If such a list must exists it should be published somehow by mono and read by mmp (so a mono fix won't require an XM update)

File: tools/mmp/driver.cs [hunk 4]
```diff
@@ -442,6 +442,29 @@ static void Main2 (string [] args)
...
+				foreach (var asm in references) {
+					List<ExclusionItem> possibleMatches = exclusionItems.Where (x => x.Name == Path.GetFileName (asm)).ToList ();
+					if (possibleMatches.Count > 0) {
+						var module = ModuleDefinition.ReadModule (asm);
+						if (possibleMatches.Any (x => x.GUID == module.Mvid))
+							referencesToFix.Add (asm);
+					}
+				}
+				if (referencesToFix.Count > 0) {
+					string facadeRoot = string.Format ("/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/{0}/Facades/", IsUnifiedMobile ? "Xamarin.Mac" : "4.5");
```
Comment (radical) [inline tools/mmp/driver.cs]: Should this use `.../Current/...` or can it the get the path in a better manner?

File: tools/mmp/driver.cs [hunk 5]
```diff
@@ -442,6 +442,29 @@ static void Main2 (string [] args)
 
 			}
 
+			if (IsUnified) {
```
Comment (spouliot) [inline tools/mmp/driver.cs]: We should exclude Modern/Mobile as we do not want to break a profile by breaking some.
The alternative would be testing with Modern/Mobile - but that seems more resource intensive (and not as safe)

File: tools/mmp/driver.cs [hunk 6]
```diff
@@ -442,6 +442,29 @@ static void Main2 (string [] args)
...
+				List<ExclusionItem> exclusionItems = ReadExclusionList ();
+				List<string> referencesToFix = new List<string> ();
+
+				foreach (var asm in references) {
+					List<ExclusionItem> possibleMatches = exclusionItems.Where (x => x.Name == Path.GetFileName (asm)).ToList ();
+					if (possibleMatches.Count > 0) {
+						var module = ModuleDefinition.ReadModule (asm);
```
Comment (spouliot) [inline tools/mmp/driver.cs]: that's done outside the resolver :| so it kills all caching :(
even more of a reason to exclude modern/mobile

File: tools/mmp/generateExclusionList.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Comment (rolfbjarne) [inline tools/mmp/generateExclusionList.sh]: Pass -e so errors aren't ignored:

```shell
#!/bin/bash -e
```

File: tools/mmp/driver.cs (hunk 4 above)
Comment (rolfbjarne) [inline tools/mmp/driver.cs]: Yes, this should not hardcode the path, I think you can use `MMPDirectory` and derive from there.

File: tools/mmp/driver.cs (hunk 1 above)
Comment (chamons) [inline tools/mmp/driver.cs]: The gist -> patch process apparently killed my whitespace :(

File: tools/mmp/Makefile (hunk 3 above)
Comment (chamons) [inline tools/mmp/Makefile]: The _original_ plan was just that. Mono publish and install a list we can consume.

And I 100% agree. This is not going in master.

**Comment by chamons** [comment]
I am having trouble validating this fixes things with the newest mono. 

2017-05-18 14:52:26.936 Test Roslyn App[39431:8994894] info: Denying load of problematic image /Users/donblas/Downloads/TestRoslynApp 2/TestRoslynApp/bin/Debug/Test Roslyn App.app/Contents/MonoBundle/System.ValueTuple.dll

And SVT is not in image.c

**Comment by chamons** [comment]
So the roslyn example in https://bugzilla.xamarin.com/show_bug.cgi?id=55988 uses SYSTEM mono, which means you need both a XM with this fix AND a mono after:

4dc09dc922210a68b272d009b6b5655d4e79f66d
4c98dc200646d45e3ce529157f760c18d4d73337
82d42e61b13ac74402543c91b66f848c8eaf042e

in mono 2017-02 to get it working.

**Pull request merged**